sourcegraph/sourcegraph
# GlobalDefs: add filters for Languages, NotLanguages, Kinds, NotKinds

- No user facing change. This is a preface to a user-facing change that will be made by @attfarhan 
- Add an API to filter search results based on: `Languages` / `NotLanguages` (`Go`, `Java`, etc), and `Kinds` / `NotKinds` (`func`, `var`, etc).

In order to migrate to the new SQL scheme, users will need to run the following in a `psql` prompt:

```
-- add the new global_defs.language ID column
ALTER TABLE global_defs ADD COLUMN language smallint;
```

Additionally, only newly built/indexed repositories will be sortable by language. Repositories can be manually re-indexed by running `src search --refresh=github.com/foo/bar` for each repository (adding a reindexing command for _all repositories_ is left as a TODO, which I will attend to this week).

##### Reviewer tasks

- I'm confident the migration SQL produces the same table schema that we desire, that said, adding the new `language` column to the `global_defs` table will lock the table.. I don't have a measurement for how long. Can anyone confirm how long this will take (since we will be without DB for that time period)?

- [x] PERF: reviewer approves of the site perf impact of this change
  - The new languages column should not have a serious perf impact, given it is a `int16` / `smallint` type.
- [x] UNIT-TEST: reviewer approves the unit tests (or the justified omission thereof)
  - Unit tests not worth the cost at this point.

##### Test plan

- Testing: on local machine:
  - Spin up a `master` revision SG instance.
  - Run the above migration code.
  - Spin up a `sg/global-defs-filters` SG instance.
  - Confirm that global search continues to work.
- Final testing+deployment: on staging, then in prod:
  - Pin the deployed version to `master`
  - Run the above migration code.
  - Deploy the new `master` (this branch).
